### DISCUSSION TITLE
Frontend development in Sylius (non-JS)

### DISCUSSION THREAD

**Discussion by vvasiloi** [Sylius/Sylius#13511] [Ideas]
A discussion complementary to https://github.com/Sylius/Sylius/issues/13510.
Please read the issue for details.

**Comment by Prometee** [comment]
Main complain would be to have more atomic `sylius_ui` events inside each Sylius template to allow plugins developers to have more easy way to add new contents. This can avoid many diff migration process I guess.

**Comment by Roshyo** [comment]
One of the main issues I've seen is regarding folder structure.

Sylius' way it to put inside templates/bundles/xxxBundle/ every overridden template. And inside templates/xxx any custom-made template. (see screen)


But that way means that any FE dev that needs to work on the project needs to be fully aware of what template is Sylius (or Plugin) made, and what template is home made. Even after figuring this out, it means also having to switch between 2 folders to gather all the templates when working.
![image (6)](https://user-images.githubusercontent.com/9363039/150113648-ee5e4ed9-73e3-4ac8-bcdb-cecaede5d850.png)


An alternative approach would be to use this kind of architecture:
![image](https://user-images.githubusercontent.com/9363039/150113810-d4050684-8188-45fd-9720-8357934eae29.png)

Where every template is placed inside the bundle folder. So if one wants to add a new template to the checkout page, they will need to put it inside bundles/SyliusShopBundle/Checkout/xxx.html.twig

That way, there is only one folder to browse when working, which makes it easier for FE team.

But it has one downside of mixing up templates, making the migration process harder.

**Reply by mbabker** [reply]
> Sylius' way it to put inside templates/bundles/xxxBundle/ every overridden template. And inside templates/xxx any custom-made template. (see screen)

That's the Symfony convention for overriding templates from bundles, Sylius just adopted it (and uses it in the theme bundle as well).

The issue with trying to put all overrides into one folder (be it a bundle folder like `templates/bundles/SyliusShopBundle` or some convention-based path like `templates/overrides` is you then have to mess with the Twig filesystem loader to look in that one path for every template and you then have to make sure that every template path for every installed bundle is globally unique (as in you couldn't have two bundles with an `Admin/Product/show.html.twig` template, how would the system know what is the right bundle to override in this case?).

**Reply by Roshyo** [reply]
The idea is not to change the path of templates or anything like that. But to have all `Shop` related templates inside `bundles/SyliusShopBundle` and all `Admin` related templates inside `bundles/SyliusAdminBundle`
And for home-made templates, the idea to treat them as Sylius native ones. So if you create a `Admin/Brand/show.html.twig` you'll call it `include '@SyliusAdmin/Brand/show.html.twig` and place it in `bundles/SyliusAdminBundle/Brand/show.html.twig`

**Reply by mbabker** [reply]
Bundles can't put templates in other bundle paths without manipulating the Twig filesystem loader in a compiler pass (Symfony only supports the template override path in the root application, it doesn't support bundles overriding bundle templates), and that is going to break the convention used by Symfony applications in general (which makes their templating documentation essentially useless in the context of Sylius).  And again, you'd run into issues if different bundles have templates with the same name.

Both SyliusShopBundle and BitBagSyliusElasticsearchPlugin provide what would be a `@SyliusShop/Product/index.html.twig` template with your suggestion.  Which one should be resolved if both bundles are publishing templates to the `@SyliusShop` Twig namespace?  What happens if you install a third bundle which provides that template?  I think trying to publish every template into a `@SyliusAdmin` or `@SyliusShop` Twig namespace would just create more problems than it fixes.

**Reply by Roshyo** [reply]
This idea is not regarding where plugins should put their templates. That does not change. Here it is only regarding final projects. Actual shops that will be running.

**Reply by mbabker** [reply]
If plugins aren’t all publishing their templates to the same `@SyliusShop` Twig namespace, then this can’t work.  The entire reason the override feature works now is because TwigBundle prepends the override path to every Twig namespace it registers.  If Sylius did something where it forced the `@SyliusShop` Twig namespace to be checked first for every template in every namespace, you still run into the issue of then requiring every template in every plugin to have a unique name otherwise you would not be able to distinguish a `Product/index.html.twig` template from one of any of the Twig namespace paths.

**Comment by 4c0n** [comment]
My biggest issue is the use of Semantic-UI. It's really not great.
Also having 2 different build systems (gulp or webpack encore) is not great. Although I believe there are plans to drop gulp.

**Reply by vvasiloi** [reply]
Replacing Semantic-UI is a huge effort, especially when the in the long term Sylius is moving towards a headless solution.
Do you have an opinion about using Fomantic UI as a temporary solution?
For dropping Gulp we have to revive and merge this PR: https://github.com/Sylius/Sylius/pull/12043

**Reply by 4c0n** [reply]
It would be better than Semantic-UI, as at least maintenance is done there. Someone was recently posting about that on Slack as well. I have tried doing that in the past, but it wasn't perfect, but that's a few years ago and things might have become better now.

As for the buildsystem I've been using Webpack + encore for quite some time now, it should be as easy as just dropping the Gulpfile, updating the dependencies and the webpack configuration. I'm not sure about updating those paths in the templates as is done in the PR, there should be an easier way somehow, but if that is what it takes, then it would still be worth the effort.

**Reply by vvasiloi** [reply]
I replaced Semantic-UI with Fomantic-UI in a project just by changing the dependencies, but that was a few years ago and I used the version of Fomantic-UI that was a drop-in replacement, fully compatible.
But now,  if we migrate, we should also upgrade and use the latest stable version.

I'm also using Webpack with Encore for a bit over 3 years already. The first migration was a bit tricky, but it got easier.
One of the main reasons that PR is stale is it needs an clear, tested upgrade path.
I said I would do that, but I didn't and have no plans for it in the next 3-4 weeks.

**Reply by 4c0n** [reply]
That's probably where I went wrong, not using the drop-in replacement version. It was like 2 years ago or something, but ended up using the Bootstrap theme instead as it was easier. My original plan was to help evolve/mature the bootstrap theme far enough that we could use it as a replacement for Semantic-UI, but it would cause a lot of BC issues, especially for all the plugins out there.

**Reply by vvasiloi** [reply]
It's not a bad idea. I also used the Bootstrap Theme a few times and was a smooth DX.

**Reply by mbabker** [reply]
The sooner Gulp is dropped from Sylius core and all of the template repos, the better.  The default Gulp build in the plugin skeleton [is actually broken](https://github.com/Sylius/PluginSkeleton/issues/291).

IMO, Semantic-UI is fine for the admin UI.  But, I think you'd have to move a lot of stuff out of SyliusUiBundle to keep using it for the admin theme, otherwise, the admin design decisions will always leak outside of the admin UI.  Since Pagerfanta is my "product" right now, I'll point out that one of those pain points is with the `@SyliusUi/Macro/pagination.html.twig` template and its "simple" pagination macro.  With how templates are loaded, you can't override this macro only for the frontend UI, so any changes here would impact anything using the UiBundle's layouts, meaning the only practical option to change the pagination display for the frontend is to override the `@SyliusShop/Product/Index/_main.html.twig` template and not use the macro (which, you are most likely overriding this template anyway if you aren't using Semantic-UI for the frontend, but the fact you can't reuse the pagination helper would be a pain point).

And, even if there are plans for the main Sylius distro to go headless, I do think a lot of consumers would appreciate SyliusShopBundle continuing to exist in some form.  I know for my company's Sylius builds, the shop bundle is more than capable of handling their needs and we might have serious pushback if we had to rebuild their frontends with a different set of tooling if the "fully featured" frontend were to stop being supported.

**Reply by vvasiloi** [reply]
AFAIK there are no plans to discontinue the shop UI.
If anything, this whole discussion aims to make customizing it and developing new UIs easier.

**Reply by mbabker** [reply]
I've seen a lot of chatter about moving toward headless, so I didn't know if dropping ShopBundle was also being considered.  As long as it'll continue to exist (even with the headless plans), then that's good news to me.

**Comment by 4c0n** [comment]
For the e-mails it would be great to have inky support: https://get.foundation/emails/docs/inky.html

**Reply by vvasiloi** [reply]
Looks good! Do you know other similar solutions?

**Reply by 4c0n** [reply]
There is also this one: https://mjml.io/
Mailchimp also has it's own language, but I'm not sure if there is also a PHP compatible engine that can integrated with twig for example.

**Reply by loevgaard** [reply]
Don't you feel like the priority for emails should be to implement Symfony Mailer and then use the tooling that comes alongside that for templates? (I don't know if Inky is part of the tooling, just throwing my thoughts out here :))

**Reply by vvasiloi** [reply]
Those two don't block each other and support for Symfony Mailer is alreayd WIP.
Also, @loic425 is going on a killing spree and Sylius Mailer is one of the victims.

**Reply by loevgaard** [reply]
Great!

**Reply by 4c0n** [reply]
Yeah it's easier to do with Symfony Mailer though, so would definitely make sense to have that sorted first, Inky can be part of the tooling for sure: https://symfony.com/doc/current/mailer.html#inky-email-templating-language

**Reply by Roshyo** [reply]
My opinion here is that since emails are not a e-commerce top priority, we should just propose the Symfony mailer implementation, and then make sure we allow everyone to use w/e they want.

Some users won't even use native Sylius emails, some will prefer mjml, some inky, and even some other custom-made solution.

**Reply by vvasiloi** [reply]
I think that transactional email are very important in e-commerce and we should not neglect this aspect.
If we don't provide a built-in solution, then we need strong integration options, cookbooks / guides, etc.

**Reply by 4c0n** [reply]
I totally agree that the e-mails are quite important! The first thing you want to do after seeing the text only e-mails that are sent from Sylius Standard is to customize them, but right now there is no nice way to do that. Inky is somewhat endorsed by Symfony (hence the mention in the Symfony docs and the package by Fabien Potencier), so it would make a lot of sense to have it available in Sylius IMO and by default make them look a bit more impressive.

**Comment by 4c0n** [comment]
I kind of feel that the templates shouldn't be part of the bundles at all. They're really only used in a Sylius Standard based setup, so why not just have them be part of that? Or are people actually using the bundles and the templates separately?
If templates need to be added for plugins for example, it could be done using Flex.

**Reply by vvasiloi** [reply]
That's also an interesting idea that aligns with the headless direction of Sylius.
Most of the templates are in Shop, Admin and UI bundles.
There are a few in the Core bundle and some in the other bundles, but those are not relevant for the front-end development.
The UI bundle exists just to reuse code between the Shop and the Admin.
This separation should make it easier to extract the front-end stuff to separate packages.

**Reply by 4c0n** [reply]
The re-usability of the UI bundle can also be a pain sometimes, for example when wanting to override a form theme for only the shop or admin side. The only way to it is to override that part of the UI bundle in the shop or admin bundle, which then renders it kind pointless.
It's great when no customizations are needed, but that is a rare case in my experience.

**Reply by vvasiloi** [reply]
I know there is a dedicated template for form themes in both Admin and Shop, that extend the one from UI.
So you **can** override it only for Admin or Shop, but there are probably some exceptions from this rule.
Would be great if you could point out these cases.

**Reply by 4c0n** [reply]
It's been a long time, but if I recall correctly the grid filters are an example of that:
https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/UiBundle/Resources/views/Grid/Filter/money.html.twig

**Reply by mbabker** [reply]
> I kind of feel that the templates shouldn't be part of the bundles at all. They're really only used in a Sylius Standard based setup, so why not just have them be part of that?

That really only works well if Sylius stops providing any UI out-of-the-box of any kind (public-facing frontend or staff-use admin).  I would expect that a fresh install of Sylius using the standard template would still have a functional admin UI, so the admin templates definitely should live alongside the AdminBundle.

Like I suggested in another comment, I think the ShopBundle should still exist in some form, even if it's not part of the out-of-the-box install (or even a first-party bundle).  I do think a lot of consumers would appreciate having some kind of ready-to-go frontend package somewhere (and there could be options for this, and not all of them have to be first-party solutions; i.e. there could be the traditional ShopBundle or point folks toward https://github.com/BitBagCommerce/SyliusVueStorefrontPlugin for a Vue Storefront integration), and I think those integrations would work better if they are maintained solutions and not recipes that get applied to an application one time.

**Reply by 4c0n** [reply]
It would still provide that. However in the form of either a theme or based on the standard implementation.
I don't see the problem with a decoupled frontend at all. If anything it would be an improvement and save a lot of issues with overriding/customizing.

**Reply by vvasiloi** [reply]
That looks similar to what I described in [this comment](https://github.com/Sylius/Sylius/discussions/13511#discussioncomment-1999705).

**Comment by 4c0n** [comment]
In order to avoid having to write controller actions for every little bit of data that might be needed inside the event templates a long time ago I wrote this expression language support thing for the event system:

https://github.com/Sylius/Sylius/pull/11591

If anyone is interested I could revive it, otherwise let's close it. :)

**Reply by vvasiloi** [reply]
I have mixed feelings about that one.

**Comment by vvasiloi** [comment]
An idea I had for some time is to have a template skeleton without any UI frameworks.
Just like there is a skeleton for plugin, there should be one for themes.
It should have Twig templates with HTML, some essential CSS/JS and build process with Webpack and Encore.
It could incorporate some best practices, latest industry standards. accessibility, SEO, etc.
The goal is to have a clean starting point for theme and custom UI development.
We can then make a pluggable default theme out of the current shop UI.

**Comment by loevgaard** [comment]
I am very happy that this discussion is even happening - that not everybody is so focused on that headless thing :D

I like the idea of completely separating admin and shop. For me, obviously, I would love the admin to look more like 2022, but certainly not something I would prioritize. I would 100 times more enjoy if the frontend looked more like 2022 :D I am more or less on the same page as Victor regarding an 'template skeleton' or something, but I would instead vote for choosing a framework, i.e. Tailwind.

But again, big thanks for even having this discussion ❤️

**Reply by vvasiloi** [reply]
Tailwind... Yikes!

<details>
  <summary>See meme</summary>
  
![image](https://user-images.githubusercontent.com/7114562/150296805-c0924be3-5804-4554-8cd4-70cd8333bf4b.png)
  
</details>

Memes aside, if it's preprocessed in a way that doesn't leak all those utility classes to HTML then I don't mind using Tailwind.

**Reply by loevgaard** [reply]
I love tailwind so far :D And the whole idea is to have those classes directly in the HTML ;D

**Comment by 4c0n** [comment]
Since Sylius is going the API platform route, it would make sense to me base the admin on the admin that is included with API Platform.
It is based on React Admin. I know this discussion says "non-js", however I guess progressive web apps are kind of the way forward. It's really not that scary and provides a lot of cool stuff, like automatic updates using Mercure etc..

**Reply by vvasiloi** [reply]
It will make updating end user projects and plugins with admin customizations very difficult.
It will probably happen eventually, but it's not going to be soon. Maybe in Sylius 3.0.

**Reply by 4c0n** [reply]
Updating yes, but with a new major version that is what can be expected. With the speed things are currently going 3.0 seems very very very far away, hopefully there will be an option to develop an alternative admin sooner. The "competition" is solving the problem by developing an SDK: https://github.com/shopware/admin-extension-sdk

**Reply by vvasiloi** [reply]
That looks cool! Great idea!

**Reply by rimas-kudelis** [reply]
> hopefully there will be an option to develop an alternative admin sooner

I would think that's what the Admin part of the new API is for, would you not?

**Reply by 4c0n** [reply]
Not necessarily, it can also be used for integrations with for example a PIM system.

**Reply by rimas-kudelis** [reply]
Of course not necessarily, but I mean it **does** (or rather **will**, when all functionality is covered) give you an option to develop alternative admin interfaces.

**Reply by 4c0n** [reply]
Ok sure you can do that, but I kind of meant in the context of a standard admin developed/supported by the Sylius team instead of third party.

**Comment by dbalabka** [comment]
Based on my experience, the current state of Sylius the frontend development isn't satisfactory. It misses the possibility of performing simple dynamic behavior of UI. Still, building a storefront for headless e-commerce using Next/Nuxt.JS requires significant effort, especially when SEO is a requirement. 

Recently, Sensio Labs started [the Symfony UX initiative](https://symfony.com/blog/new-in-symfony-the-ux-initiative-a-new-javascript-ecosystem-for-symfony), which provides a very basic javascript ecosystem. [Symfony Turbo](https://ux.symfony.com/turbo) is the backbone of this idea, providing the speed of SPA w/o the need to write JS. However, it might be extended with [Stimulus](https://stimulus.hotwired.dev/) to create more complicated UX experiences.

It looks like a very logical enhancement for the existing templating/theming approach without the need to abandon Twig.

**Reply by wadjeroudi** [reply]
Agree on that approach same goes for the admin, should be based on symfony turbo. for example grids could be autoreloaded when an order is updated.

**Reply by ikakarov** [reply]
I work on a big project using Sylius.
 Every customer who uses the administration is unhappy because it is too basic and poor. 
 For example, we need three basic features that are present in every administration:
1. We need a media browser - that is very important when we have tons of records.
2. Dynamic grids - something like datagrid - that is crucial when you need to change or view large datasets.
3. Modal editing - something like Sonata modal edit - for example, I need to add a new tag to a product. I open the product, search for the tag, but for the creation step, I need to go to another page.

If I have some free time, I am considering starting the Symfony LAST* stack to create a simple frontend theme.
*Symfony LAST stack is:
Live Components
Asset Mapper
StimulusJS
Turbo

I believe that the poor frontend is one of the main reasons why this platform does not become one of the most widely used.

**Comment by CoderMaggie** [comment]
Hi People! What are your current thoughts on this topic regarding the general frontend updates we're experiencing in the `2.0` and `bootstrap-shop` branches? 👀 
is there anything we should take a look at in the upcoming development of the 2.x versions?

**Reply by dbalabka** [reply]
1. Headless shop theme with Vue/React support
2. Standard library of Vue/React components for e-commerce frontstore.
3. SEO optimization support
4. Optimized PageSpeed metrics
5. Considering NextJS/NuxtJS support
6. Considering Material design system support or any other: https://adele.uxpin.com/
7. Considering providing Storybook out-of-the-box: https://storybook.js.org/

**Reply by dbalabka** [reply]
8. It would be amazing to have Sylius API JS Client library to Supabase JS Client library.
